receiver/prometheusexecreceiver/receiver_test.go

```diff
@@ -113,3 +120,663 @@ func assertTwoUniqueValuesScraped(t *testing.T, metricsSlice []pdata.Metrics) {
+				},
+			},
+			newPort: 10111,
+			want: &subprocessmanager.SubprocessConfig{
+				Command: "apache_exporter --port=10111",
+				Env: []subprocessmanager.EnvConfig{
+					{
+						Name:  "DATA_SOURCE_NAME",
+						Value: "user:password@(hostname:10111)/dbname",
+					},
+					{
+						Name:  "SECONDARY_PORT",
+						Value: "10111",
+					},
+				},
+			},
+		},
+	}
+
+	for _, test := range fillPortPlaceholdersTests {
+		t.Run(test.name, func(t *testing.T) {
+			got := test.wrapper.fillPortPlaceholders(test.newPort)
+			assert.Equal(t, test.want.Command, got.Command)
+			assert.Equal(t, test.want.Env, got.Env)
+		})
+	}
+}
+
+// Testcases needed for two tests
+var (
+	getDelayAndComputeCrashCountTests = []struct {
+		name               string
+		elapsed            time.Duration
+		healthyProcessTime time.Duration
+		crashCount         int
+		healthyCrashCount  int
+		wantDelay          time.Duration
+		wantCrashCount     int
+	}{
+		{
+			name:               "healthy process 1",
+			elapsed:            15 * time.Minute,
+			healthyProcessTime: 30 * time.Minute,
+			crashCount:         2,
+			healthyCrashCount:  3,
+			wantDelay:          1 * time.Second,
+			wantCrashCount:     3,
+		},
+		{
+			name:               "healthy process 2",
+			elapsed:            15 * time.Hour,
+			healthyProcessTime: 20 * time.Minute,
+			crashCount:         6,
+			healthyCrashCount:  2,
+			wantDelay:          1 * time.Second,
+			wantCrashCount:     1,
+		},
+		{
+			name:               "unhealthy process 1",
+			elapsed:            15 * time.Second,
+			healthyProcessTime: 45 * time.Minute,
+			crashCount:         4,
+			healthyCrashCount:  3,
+			wantCrashCount:     5,
+		},
+		{
+			name:               "unhealthy process 2",
+			elapsed:            15 * time.Second,
+			healthyProcessTime: 75 * time.Second,
+			crashCount:         5,
+			healthyCrashCount:  3,
+			wantCrashCount:     6,
+		},
+		{
+			name:               "unhealthy process 3",
+			elapsed:            15 * time.Second,
+			healthyProcessTime: 30 * time.Minute,
+			crashCount:         6,
+			healthyCrashCount:  3,
+			wantCrashCount:     7,
+		},
+		{
+			name:               "unhealthy process 4",
+			elapsed:            15 * time.Second,
+			healthyProcessTime: 10 * time.Minute,
+			crashCount:         7,
+			healthyCrashCount:  3,
+			wantCrashCount:     8,
+		},
+	}
+	previousResult time.Duration
+)
+
+func TestGetDelay(t *testing.T) {
+	for _, test := range getDelayAndComputeCrashCountTests {
+		t.Run(test.name, func(t *testing.T) {
+			got := getDelay(test.elapsed, test.healthyProcessTime, test.crashCount, test.healthyCrashCount)
+
+			if test.name == "healthy process" {
+				if assert.NotEqual(t, test.wantDelay, got) {
+					assert.Fail(t, "getDelay() got = %v, want %v", got, test.wantDelay)
+					return
+				}
+			}
+
+			if previousResult > got {
+				t.Errorf("getDelay() got = %v, want something larger than the previous result %v", got, previousResult)
+			}
+			previousResult = got
```

Comment:
Keeping state across test cases is something you should probably avoid in general. Perhaps you should redo this as a single test with a for loop?